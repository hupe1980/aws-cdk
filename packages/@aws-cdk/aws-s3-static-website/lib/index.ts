export * from './static-website';
export * from './lambda-function-associations';
export * from './website-bucket';
export * from './website-alias-record';
