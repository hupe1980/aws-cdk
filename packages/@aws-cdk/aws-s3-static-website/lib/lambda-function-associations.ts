import {
  CfnDistribution,
  CloudFrontWebDistribution
} from '@aws-cdk/aws-cloudfront';
import { Construct } from '@aws-cdk/core';

export interface Association {
  /**
   * The ARN of the Lambda function
   */
  readonly functionArn: string;

  /**
   * The version of the Lambda function
   */
  readonly functionVersion: string;

  /**
   * Specifies the event type that triggers a Lambda function invocation
   */
  readonly eventType:
    | 'viewer-request'
    | 'origin-request'
    | 'origin-response'
    | 'viewer-response';
}

export interface LambdaFunctionAssociationsProps {
  /**
   * The CloudFront distribution
   */
  readonly distribution: CloudFrontWebDistribution;

  /**
   * The Lambda function associations
   */
  readonly assosiations: Association[];
}

export class LambdaFunctionAssociations extends Construct {
  constructor(
    scope: Construct,
    id: string,
    props: LambdaFunctionAssociationsProps
  ) {
    super(scope, id);

    const cfDist = props.distribution.node.findChild(
      'CFDistribution'
    ) as CfnDistribution;

    const assosiations = props.assosiations.map(assosiation => ({
      EventType: assosiation.eventType,
      LambdaFunctionARN: `${assosiation.functionArn}:${
        assosiation.functionVersion
      }`
    }));

    cfDist.addOverride(
      'Properties.DistributionConfig.DefaultCacheBehavior.LambdaFunctionAssociations',
      assosiations
    );
  }
}
