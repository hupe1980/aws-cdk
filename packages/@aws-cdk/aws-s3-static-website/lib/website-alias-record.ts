import {
  ARecord,
  HostedZone,
  IAliasRecordTarget,
  RecordTarget
} from '@aws-cdk/aws-route53';
import { Construct } from '@aws-cdk/core';

export interface WebsiteAliasRecordProps {
  /** The domain name for the site like 'example.com' */
  readonly domainName: string;

  /** Names for the records. */
  readonly recordNames: string[];

  /** Target for the alias record */
  readonly target: IAliasRecordTarget;
}

export class WebsiteAliasRecord extends Construct {
  constructor(scope: Construct, id: string, props: WebsiteAliasRecordProps) {
    super(scope, id);

    const { domainName, recordNames, target } = props;

    const zone = HostedZone.fromLookup(this, 'Zone', {
      domainName
    });

    recordNames.forEach(
      (recordName, idx): void => {
        new ARecord(this, `WebsiteAliasRecord${idx}`, {
          zone,
          recordName: `${recordName}.`,
          target: RecordTarget.fromAlias(target)
        });
      }
    );
  }
}
