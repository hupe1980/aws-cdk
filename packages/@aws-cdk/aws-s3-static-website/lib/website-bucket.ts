import {
  CfnCloudFrontOriginAccessIdentity,
  S3OriginConfig
} from '@aws-cdk/aws-cloudfront';
import { CanonicalUserPrincipal } from '@aws-cdk/aws-iam';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Construct, RemovalPolicy } from '@aws-cdk/core';
import * as path from 'path';

export interface WebsiteBucketProps {
  /**
   * Name of the bucket
   *
   * @default - Assigned by CloudFormation (recommended).
   */
  readonly bucketName?: string;

  /**
   * Policy to apply when the bucket is removed from this stack.
   *
   * @default - The bucket will be orphaned.
   */
  readonly removalPolicy?: RemovalPolicy;

  /**
   * The source from which to deploy the website
   *
   * @default - Dummy placeholder
   */
  readonly source?: string;

  /**
   * Disable website deployment
   *
   * @default - false
   */
  readonly disableUpload?: boolean;

  /**
   * The index page for the site like 'index.html'
   *
   * @default - index.html
   */
  readonly websiteIndexDocument?: string;

  /**
   *  The error page for the site like 'error.html'
   *
   *  @default - error.html
   */
  readonly websiteErrorDocument?: string;
}

export class WebsiteBucket extends Construct {
  public readonly s3OriginConfig: S3OriginConfig;

  constructor(scope: Construct, id: string, props: WebsiteBucketProps = {}) {
    super(scope, id);

    const {
      bucketName,
      removalPolicy = RemovalPolicy.RETAIN,
      disableUpload = false,
      source,
      websiteIndexDocument,
      websiteErrorDocument
    } = props;

    const bucket = new Bucket(this, 'WebsiteBucket', {
      bucketName,
      removalPolicy,
      websiteIndexDocument: websiteIndexDocument || 'index.html',
      websiteErrorDocument: websiteErrorDocument || 'error.html'
    });

    const originId = new CfnCloudFrontOriginAccessIdentity(
      this,
      'OriginAccessIdentity',
      {
        cloudFrontOriginAccessIdentityConfig: {
          comment: `CloudFront OriginAccessIdentity for ${bucket.bucketName}`
        }
      }
    );

    bucket.grantRead(
      new CanonicalUserPrincipal(originId.attrS3CanonicalUserId)
    );

    if (!disableUpload) {
      const placeHolderSource = path.join(__dirname, '..', 'website');

      new BucketDeployment(this, 'WebsiteDeployment', {
        source: Source.asset(source || placeHolderSource),
        destinationBucket: bucket,
        retainOnDelete: removalPolicy === RemovalPolicy.RETAIN
      });
    }

    this.s3OriginConfig = {
      originAccessIdentityId: originId.ref,
      s3BucketSource: bucket
    };
  }
}
