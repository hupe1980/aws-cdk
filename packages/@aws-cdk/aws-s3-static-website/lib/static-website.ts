import {
  AliasConfiguration,
  CloudFrontWebDistribution,
  CloudFrontWebDistributionProps
} from '@aws-cdk/aws-cloudfront';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { Construct } from '@aws-cdk/core';

import {
  Association,
  LambdaFunctionAssociations
} from './lambda-function-associations';
import { WebsiteAliasRecord } from './website-alias-record';
import { WebsiteBucket, WebsiteBucketProps } from './website-bucket';

export interface AliasProps extends AliasConfiguration {
  /** The domain name for the site like 'example.com' */
  readonly domainName: string;
}

export interface StaticWebsiteProps extends WebsiteBucketProps {
  /**
   * BucketConfiguration is used to configured the S3 website bucket
   *
   * @default - The website bucket is provided with default values
   */
  readonly bucketConfiguration?: WebsiteBucketProps;

  /**
   * AliasConfiguration is used to configured CloudFront to respond to requests on custom domain names.
   *
   * @default - No custom domain names are set up
   */
  readonly aliasConfiguration?: AliasProps;

  /**
   * AWS WAF WebACL to associate with this CloudFront distribution
   *
   * @default - No AWS Web Application Firewall web access control list (web ACL)
   */
  readonly webACLId?: string;
}

export class StaticWebsite extends Construct {
  private readonly distribution: CloudFrontWebDistribution;

  constructor(scope: Construct, id: string, props: StaticWebsiteProps = {}) {
    super(scope, id);

    const { aliasConfiguration, bucketConfiguration, webACLId } = props;

    const websiteBucket = new WebsiteBucket(this, 'WebsiteBucket', {
      ...bucketConfiguration
    });

    const distibutionConfig: CloudFrontWebDistributionProps = {
      webACLId,
      originConfigs: [
        {
          s3OriginSource: {
            ...websiteBucket.s3OriginConfig
          },
          behaviors: [{ isDefaultBehavior: true }]
        }
      ],
      aliasConfiguration
    };

    this.distribution = new CloudFrontWebDistribution(
      this,
      'WebSiteDistribution',
      distibutionConfig
    );

    if (aliasConfiguration) {
      new WebsiteAliasRecord(this, 'WebsiteAliasRecord', {
        domainName: aliasConfiguration.domainName,
        recordNames: aliasConfiguration.names,
        target: new CloudFrontTarget(this.distribution)
      });
    }
  }

  public addLambdaFunctionAssociation(assosiation: Association) {
    return this.addLambdaFunctionAssociations([assosiation]);
  }

  public addLambdaFunctionAssociations(assosiations: Association[]) {
    return new LambdaFunctionAssociations(this, 'LambdaFunctionAssociation', {
      distribution: this.distribution,
      assosiations
    });
  }
}
