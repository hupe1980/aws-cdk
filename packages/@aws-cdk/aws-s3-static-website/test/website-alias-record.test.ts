import { CloudFrontWebDistribution } from '@aws-cdk/aws-cloudfront';
import { HostedZone } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets';
import { Bucket } from '@aws-cdk/aws-s3';
import { Stack } from '@aws-cdk/core';
import '@cloudcomponents/jest-cdk-snapshot';
import { WebsiteAliasRecord } from '../lib';

test('default setup', (): void => {
  const stack = new Stack();

  jest.spyOn(HostedZone, 'fromLookup').mockImplementation(
    () =>
      new HostedZone(stack, 'HostedZone', {
        zoneName: 'example.com'
      })
  );

  const distribution = new CloudFrontWebDistribution(
    stack,
    'WebSiteDistribution',
    {
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: new Bucket(stack, 'DummyWebsiteBucket')
          },
          behaviors: [{ isDefaultBehavior: true }]
        }
      ]
    }
  );

  new WebsiteAliasRecord(stack, 'WebsiteAliasRecord', {
    domainName: 'example.com',
    recordNames: ['www.example.com', 'example.com'],
    target: new CloudFrontTarget(distribution)
  });

  expect(stack).toMatchCdkSnapshot();
});
