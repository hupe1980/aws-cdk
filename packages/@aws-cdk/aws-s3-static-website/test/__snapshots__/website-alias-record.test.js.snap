// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Resources": Object {
    "DummyWebsiteBucket4BEF8CBE": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "HostedZoneDB99F866": Object {
      "Properties": Object {
        "Name": "example.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "WebSiteDistributionCFDistributionFAD03684": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "CacheBehaviors": Array [],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "DummyWebsiteBucket4BEF8CBE",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {},
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "WebsiteAliasRecordWebsiteAliasRecord04277EB1F": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "WebSiteDistributionCFDistributionFAD03684",
              "DomainName",
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "www.example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "WebsiteAliasRecordWebsiteAliasRecord16ED9533E": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "WebSiteDistributionCFDistributionFAD03684",
              "DomainName",
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": Object {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
  },
}
`;
